### PR TITLE
Fix occurrence query function

### LINKED ISSUE #1108: occurrence function not working as expected
The occurence query function (#OCCURANCE) is not correct because of a couple of reasons:
1) Values are added into the JexlContext from the index as well as those from the event.  Only the values from the event should be considered in this case unless there are none.
2) The loop that tries to count the occurences is invalid in that it tries to create a map based on type, but then increments the count anyway for those that do not have a type.  Basically a somewhat random function based on the order of the values in the set.

### PR DESCRIPTION
**Pull request by lbschanno**
The occurrence function currently does not count occurrences correctly.
Add the ability to track when a field value originated from the index
vs. the event via its Attribute. Modify the occurrence function to only
return the total number of values from the event, unless there are none.
In this case, return the total number of values from the index.

Fixes #1108

### CODE REVIEW

File: warehouse/query-core/src/main/java/datawave/query/attributes/Attribute.java [hunk 1]
```diff
@@ -31,6 +33,7 @@
      */
     protected Key metadata = null;
     protected boolean toKeep = true; // a flag denoting whether this attribute is to be kept in the returned results (transient or not)
+    protected boolean fromIndex; // Assume attributes are from the index unless specified otherwise.
```
Comment (ivakegg) [inline warehouse/query-core/src/main/java/datawave/query/attributes/Attribute.java]: If this comment is correct, then should it be initialized to true?

File: warehouse/query-core/src/main/java/datawave/query/iterator/aggregation/DocumentData.java [hunk 2]
```diff
@@ -52,6 +55,14 @@ public void setData(List<Entry<Key,Value>> data) {
         this.data = data;
     }
     
+    public boolean isFromIndex() {
+        return fromIndex;
+    }
+    
+    public void setFromIndex(boolean fromIndex) {
```
Comment (ivakegg) [inline warehouse/query-core/src/main/java/datawave/query/iterator/aggregation/DocumentData.java]: Should this actually be mutable?

File: warehouse/query-core/src/main/java/datawave/query/predicate/ValueToAttributes.java [hunk 3]
```diff
@@ -52,14 +52,19 @@
     
     private EventDataQueryFilter attrFilter;
     
-    public ValueToAttributes(CompositeMetadata compositeMetadata, TypeMetadata typeMetadata, EventDataQueryFilter attrFilter, MarkingFunctions markingFunctions) {
+    // Whether the value is from the index
+    private boolean fromIndex;
```
Comment (ivakegg) [inline warehouse/query-core/src/main/java/datawave/query/predicate/ValueToAttributes.java]: final?

**Review by ivakegg — CHANGES_REQUESTED** [review]

File: warehouse/query-core/src/main/java/datawave/query/attributes/Attribute.java (hunk 1 above)
Comment (lbschanno) [reply warehouse/query-core/src/main/java/datawave/query/attributes/Attribute.java]: It should be. Fixed.

File: warehouse/query-core/src/main/java/datawave/query/iterator/aggregation/DocumentData.java (hunk 2 above)
Comment (lbschanno) [reply warehouse/query-core/src/main/java/datawave/query/iterator/aggregation/DocumentData.java]: It should not be. Good catch. Fixed.

File: warehouse/query-core/src/main/java/datawave/query/predicate/ValueToAttributes.java (hunk 3 above)
Comment (lbschanno) [reply warehouse/query-core/src/main/java/datawave/query/predicate/ValueToAttributes.java]: Fixed.

**Review by ivakegg — APPROVED** [review]

File: warehouse/query-core/src/main/java/datawave/query/jexl/functions/EvaluationPhaseFilterFunctions.java [hunk 4]
```diff
@@ -58,32 +56,30 @@ public static boolean occurrence(Iterable<?> values, int count) {
...
     
     private static int getSizeOf(Iterable<?> iterable) {
-        Map<String,Integer> countMap = Maps.newHashMap();
-        int count = 0;
         if (iterable != null) {
-            // additional datatypes may have been added to this field to assist with evaluation and for use as
-            // a 'marker' for some other condition. We care only about how many were in the original so
-            // bin them all up to count what was really there in the beginning
+            int sourcedFromEvent = 0;
+            int sourcedFromIndex = 0;
```
Comment (ivakegg) [inline warehouse/query-core/src/main/java/datawave/query/jexl/functions/EvaluationPhaseFilterFunctions.java:62]: I think I need a third category here for anything that is not a ValueTuple.  I am not sure this is needed but it seems odd to pass in an iterable of non ValueTuple objects and come back with a size of 0.  So if sourcedFromEvent is non-zero return that, else if sourceFromIndex is non-zero then return that , else return the number in the 3rd category..

**Review by ivakegg — CHANGES_REQUESTED** [review]

File: warehouse/query-core/src/main/java/datawave/query/jexl/functions/EvaluationPhaseFilterFunctions.java (hunk 4 above)
Comment (lbschanno) [reply warehouse/query-core/src/main/java/datawave/query/jexl/functions/EvaluationPhaseFilterFunctions.java:62]: This is now added.

**Review by ivakegg — DISMISSED** [review]

File: warehouse/query-core/src/main/java/datawave/query/attributes/Attribute.java [hunk 5]
```diff
@@ -15,13 +15,15 @@
 import org.apache.hadoop.io.WritableComparable;
 import org.apache.hadoop.io.WritableUtils;
 import org.apache.log4j.Logger;
+import org.w3c.dom.Attr;
```
Comment (apmoriarty) [inline warehouse/query-core/src/main/java/datawave/query/attributes/Attribute.java]: I don't think is is needed

**Review by apmoriarty — DISMISSED** [review]
LGTM

File: warehouse/query-core/src/main/java/datawave/query/attributes/Attribute.java (hunk 5 above)
Comment (lbschanno) [reply warehouse/query-core/src/main/java/datawave/query/attributes/Attribute.java]: Good catch, fixed!

**Review by apmoriarty — APPROVED** [review]
LGTM

**Review by ivakegg — APPROVED** [review]

**Pull request merged**